devops/azure-pipelines.yml

```diff
@@ -0,0 +1,97 @@
+# This pipeline does not implement the analysis template from the 
+# project-registry-devops repository because there was an issue with
+# analysis of an Angular build when checking out multiple repos.
+
+# This pipeline will trigger on commits or pull-requests made to the branch-dev
+# or main branches.  Note that full static analysis with SonarCloud is only 
+# performed on "long-lived" branches, which are designated as 'main' and any 
+# branch name prefixed with 'branch-'.  Minimal analysis is still performed on
+# "short-lived" branches and special analyisis is performed on pull-requests
+trigger:
+- branch-dev
+- main
+pr:
+- branch-dev
+- main
+
+# These variable groups are required for sonarCloud and Discord secrets to be
+# available for analysis tasks
+# The contents of these groups that are used here are:
+#  project-registry-sonarcloud: stores sonarcloud variables/secrets
+#   - sonarKeyPrefix: the common prefix string for the sonar keys in the project
+#       (requires a sonarCloud project key)
+#   - sonarNamePrefix: the common prefix string for the project names
+#   - sonarOrg: the sonar organization
+#   - sonarUrlPrefix: the common prefix string for the url for the results
+#  project-registry-discord: stores discord variables/secrets
+#   - discordAnalysisChannel: the output channel for discord analysis info
+#   - discordAnalysisKey: the webhook key associated with the above channel
+variables:
+- group: project-registry-sonarcloud
+- group: project-registry-discord
+- name: microservice
+  value: "frontend"
+- name: coveragePath
+  value: "./p3-frontend/**"
+- name: workDir
+  value: "./p3-frontend/"
+
+pool:
+  vmImage: ubuntu-latest
+
+stages:
+- stage: build
+  jobs:
+  - job: analysis
+    steps:
+    - checkout: self
+
+    - task: SonarCloudPrepare@1
+      inputs:
+        SonarCloud: 'SonarCloud'
+        organization: $(sonarOrg)
+        scannerMode: 'CLI'
+        configMode: 'manual'
+        cliProjectKey: $(sonarKeyPrefix)${{ variables.microservice }}
+        cliSources: './${{ variables.workDir }}'
+      displayName: Prepare Analysis for ${{ variables.microservice }} --> Task
```

Comment:
ah, i remembered something. you want some extra properties for an angular analysis. for example, as it is, it's expecting code coverage on the test files. i'll edit this comment with more info in a sec.

Reply:
I tried to get that working and it but it didn't appear to work, so I removed it.  If you have some documentation for me I would welcome it!

Reply:
i'll have to do a little research myself to get it properly right, i'll let you know

Reply:
@jpbulloch5 i think [this](https://github.com/2006-jun15-net/nick-project2-kitchen-site/blob/master/azure-pipelines.yml#L44) is pretty much what we want. it's a little old but it should still be good.

Reply:
I applied it to the pipeline, but I don't think it worked:
The logs include a line that says:
`WARN: No coverage information will be saved because all LCOV files cannot be found.`
I looked through the directories and found no files with 'lcov' in the filename, but I am not sure if that's what it meant.

Reply:
right... when ng test is run, it needs the --code-coverage option, as shown in that linked pipeline. you can add the coverage option at any one of several layers of abstraction - in the pipeline yml, in the npm script, and in the karma.config.